mlflow/store/artifact/databricks_artifact_repo.py

```diff
@@ -231,6 +240,71 @@ def _azure_upload_file(self, credentials, local_file, artifact_path):
         except Exception as err:
             raise MlflowException(err)
 
+    def _retryable_adls_function(self, func, artifact_path, **kwargs):
+        # Attempt to call the passed function.  Retry if the credentials have expired
+        try:
+            func(**kwargs)
+        except requests.HTTPError as e:
+            if e.response.status_code in [403]:
+                _logger.info(
+                    "Failed to authorize ADLS operation, possibly due "
+                    "to credential expiration. Refreshing credentials and trying again..."
+                )
+                new_credentials = self._get_write_credential_infos(
+                    run_id=self.run_id, paths=[artifact_path]
+                )[0]
+                kwargs['sas_url'] = new_credentials.signed_uri
+                put_adls_file_creation(**kwargs)
+            else:
+                raise e
+
+    def _azure_adls_gen2_upload_file(self, credentials, local_file, artifact_path):
+        """
+        Uploads a file to a given Azure storage location using the ADLS gen2 API.
+        """
+        try:
+            headers = self._extract_headers_from_credentials(credentials.headers)
+
+            # try to create the file
+            self._retryable_adls_function(
+                func=put_adls_file_creation,
+                artifact_path=artifact_path,
+                sas_url=credentials.signed_uri,
+                headers=headers
+            )
+
+            # next try to append the file
+            cursor_position = 0
+            is_first_chunk = True
+            use_single_part_upload = False
+            for chunk in yield_file_in_chunks(local_file, _AZURE_MAX_BLOCK_CHUNK_SIZE):
+                cur_chunk_size = len(chunk)
+                if is_first_chunk and cur_chunk_size < _AZURE_MAX_BLOCK_CHUNK_SIZE:
+                    use_single_part_upload = True
+                self._retryable_adls_function(
+                    func=patch_adls_file_upload,
+                    artifact_path=artifact_path,
+                    sas_url=credentials.signed_uri,
+                    data=chunk,
+                    position=cursor_position,
+                    headers=headers,
+                    is_single=use_single_part_upload,
+                )
+                is_first_chunk = False
```

Comment:
```suggestion
            use_single_part_upload = False
            for idx, chunk in enumerate(yield_file_in_chunks(local_file, _AZURE_MAX_BLOCK_CHUNK_SIZE):
                cur_chunk_size = len(chunk)
                if idx == 0 and cur_chunk_size < _AZURE_MAX_BLOCK_CHUNK_SIZE:
                    use_single_part_upload = True
                self._retryable_adls_function(
                    func=patch_adls_file_upload,
                    artifact_path=artifact_path,
                    sas_url=credentials.signed_uri,
                    data=chunk,
                    position=cursor_position,
                    headers=headers,
                    is_single=use_single_part_upload,
                )
```

Can we use `enumerate` and remove `is_first_chunk`?

Reply:
Will change.